### DISCUSSION TITLE
How to deal with stores being empty at creation

### DISCUSSION THREAD

**Discussion by danrspencer** [kube-rs/kube#1124] [Q&A]
When we first started using stores we encountered a bug because there can be a race condition if you just create one and give it to a Controller. If the Controller reconciles before the Store has received its first event then the Store will be empty; which can cause incorrect behaviour in your reconciliation. To solve this we've created an async wrapper function around creating a store that waits for the first event before returning the store. 

Is there a better / more idiomatic KubeRS way of solving this?

**Comment by nightkr** [comment]
So this is going to depend on your use-case, yeah... For the simple use-case of "one store for the main resource being reconciled" then this doesn't matter, since the controller will populate the store before calling into your reconciler. For 1:1 relations to secondary types it might also be ok (assuming that the secondary object is mandatory for reconciliation), since the reconciler will be called again when the secondary object is discovered (assuming correctly configured watches).

Where it becomes troublesome is 1:n and 1:(0-1) relationships, since this means that `store.get().is_none()` means "object may not exist" rather than "object does not exist".

I ended up writing a helper type for this (https://github.com/stackabletech/commons-operator/blob/6a398e08e80b4810e289162fe555fa3945d6eeb1/rust/operator-binary/src/utils/delayed_init.rs), but it's definitely pretty ugly.

There are a few options here.. we could allow users to detect the condition by changing `Store::get` to return a `Result<K, GetError>` where `enum GetError { DoesNotExist, NotInitedYet, }`, but I'm not sure that would help that often (since you can still only return an error at that point, or poll yourself). Maybe the better option would be to add a `Store::wait_for_init() -> impl Future<Output = ()>` method, but that feels like a footgun if people don't bother calling it... (not to mention the risk of deadlocks)

**Reply by danrspencer** [reply]
Cheers, your helper is pretty much identical in purpose to how we're starting our stores. Glad to see we're not the only ones struggling with this.
